dashboard/test/seeding/lesson_seeding_test.rb

```diff
@@ -0,0 +1,146 @@
+require 'test_helper'
+
+class LessonSeedingTest < ActiveSupport::TestCase
+  test 'serialize script lessons' do
+    script = create_script_with_lessons(num_lesson_groups: 2, num_lessons_per_group: 2)
+    # Set names / keys for consistent results regardless of DB state
+    script.name = 'bogus-script-1'
+    script.lesson_groups.each_with_index do |lg, i|
+      lg.key = "Bogus Lesson Group #{i + 1}"
+      lg.save!
+    end
+    script.lessons.each_with_index do |l, i|
+      l.key = "Bogus-Lesson-#{i + 1}"
+      l.save!
+    end
+
+    # Set some properties to serialize
+    lesson1, lesson2, lesson3, _ = script.lesson_groups.map(&:lessons).flatten
+    lesson1.overview = 'overview 1'
+    lesson2.visible_after = 'visible_after 2'
+    lesson3.overview = 'overview 3'
+    lesson3.visible_after = 'visible_after 3'
+
+    script.save!
+
+    expected = <<~JSON
+      {
+        "name": "bogus-script-1",
+        "lesson_groups": [
+          {
+            "key": "Bogus Lesson Group 1",
+            "lessons": [
+              {
+                "key": "Bogus-Lesson-1",
+                "properties": {
+                  "overview": "overview 1"
+                }
+              },
+              {
+                "key": "Bogus-Lesson-2",
+                "properties": {
+                  "visible_after": "visible_after 2"
+                }
+              }
+            ]
+          },
+          {
+            "key": "Bogus Lesson Group 2",
+            "lessons": [
+              {
+                "key": "Bogus-Lesson-3",
+                "properties": {
+                  "overview": "overview 3",
+                  "visible_after": "visible_after 3"
+                }
+              },
+              {
+                "key": "Bogus-Lesson-4",
+                "properties": {
+                }
+              }
+            ]
+          }
+        ]
+      }
+    JSON
+
+    assert_equal expected.strip, LessonSeeding.serialize_lessons(script)
+  end
+
+  test 'seeding sets a new property' do
+    script = create_script_with_lessons(num_lesson_groups: 1, num_lessons_per_group: 1)
```

Comment:
in the case where you only need 1 lesson group and 1 lesson, a pattern we sometimes use in factories is that when you create a lesson, a lesson group and script are automatically created too. I could be mistaken but it might just require changing `script` to `lesson_group` here: https://github.com/code-dot-org/code-dot-org/blob/e9d11735f0a2a56f7d66dd84fd505668b703e1b6/dashboard/test/factories/factories.rb#L778-L781

then maybe:
```
lesson = create :lesson
lesson_group = lesson.lesson_group
script = lesson_group.script
```

this combined with the `with_lesson_groups` / `with_lessons` idea looks like it would meet all your current testing needs and eliminate the need for this method.

Reply:
I think factories with associations are a little trickier than that, unfortunately. Just changing `script` to `lesson_group` gets us:

```
[1] pry(#<LessonSeedingTest>)> lesson = create :lesson
NoMethodError: undefined method `lessons' for nil:NilClass
from /Users/winterdong/CDO/code-dot-org/dashboard/test/factories/factories.rb:784:in `block (3 levels) in <top (required)>'
```

And adding both `script` and `lesson_group` gets us two scripts created:

```
[4] pry(#<LessonSeedingTest>)> lesson.lesson_group.script
=> #<Script:0x00007fb6054d1f80
 id: 12927,
 name: "bogus-script-2",
 created_at: Sat, 12 Sep 2020 01:07:08 UTC +00:00,
 updated_at: Sat, 12 Sep 2020 01:07:08 UTC +00:00,
 wrapup_video_id: nil,
 hidden: false,
 user_id: nil,
 login_required: false,
 properties: {},
 new_name: nil,
 family_name: nil>
[5] pry(#<LessonSeedingTest>)> lesson.script
=> #<Script:0x00007fb604c21958
 id: 12926,
 name: "bogus-script-1",
 created_at: Sat, 12 Sep 2020 01:07:08 UTC +00:00,
 updated_at: Sat, 12 Sep 2020 01:07:08 UTC +00:00,
 wrapup_video_id: nil,
 hidden: false,
 user_id: nil,
 login_required: false,
 properties: {},
 new_name: nil,
 family_name: nil>
```

I'm sure there's a way to do it, but I don't feel like fiddling with this for this PR, unless you feel strongly about it.

Reply:
no that's fine, factory additions would be great if they just worked as i had commented, but no need to block this PR on it since they don't seem to work that way.